packages/server/lib/services/activity.service.ts

```diff
@@ -0,0 +1,124 @@
+    await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).where({ id }).update({
+        session_id
+    });
+}
+
+export async function updateSuccess(id: number, success: boolean): Promise<void> {
+    await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).where({ id }).update({
+        success
+    });
+}
+
+export async function updateEndpoint(id: number, endpoint: string): Promise<void> {
+    await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).where({ id }).update({
+        endpoint
+    });
+}
+
+export async function updateAction(id: number, action: LogAction): Promise<void> {
+    await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).where({ id }).update({
+        action
+    });
+}
+
+export async function createActivityLogMessage(logMessage: ActivityLogMessage): Promise<boolean> {
+    logger.log(logMessage.level as string, logMessage.content);
+
+    try {
+        const result: void | Pick<ActivityLogMessage, 'id'> = await db.knex
+            .withSchema(db.schema())
+            .from<ActivityLogMessage>(activityLogMessageTableName)
+            .insert(logMessage, ['id']);
+
+        if (Array.isArray(result) && result.length === 1 && result[0] !== null && 'id' in result[0]) {
+            return true;
+        }
+    } catch (e) {
+        console.log(e);
+    }
+
+    return false;
+}
+
+export async function addEndTime(activity_log_id: number): Promise<void> {
+    await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).where({ id: activity_log_id }).update({
+        end: Date.now()
+    });
+}
+
+export async function createActivityLogMessageAndEnd(logMessage: ActivityLogMessage): Promise<void> {
+    await createActivityLogMessage(logMessage);
+    await addEndTime(logMessage.activity_log_id);
+}
+
+export async function findActivityLogBySession(session_id: string): Promise<number | null> {
+    const result = await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).select('id').where({ session_id });
+
+    if (!result || result.length == 0 || !result[0]) {
+        return null;
+    }
+
+    return result[0].id;
+}
+
+export async function getLogsByAccount(account_id: number): Promise<ActivityLog[]> {
+    const logs = await db.knex.withSchema(db.schema()).from<ActivityLog>(activityLogTableName).select('*').where({ account_id }).orderBy('id', 'desc');
+
+    if (!logs || logs.length == 0 || !logs[0]) {
+        return [];
+    }
+    const logsWithMessages = await Promise.all(
+        logs.map(async (log) => {
+            const messages = await db.knex
+                .withSchema(db.schema())
+                .from<ActivityLogMessage>(activityLogMessageTableName)
+                .select('*')
+                .where({ activity_log_id: log.id as number });
+            log.messages = messages;
+            return log;
+        })
+    );
```

Comment:
Is there any way to use `join` here instead of query messages for each specific type or batch the calls? I'm generally afraid of two things here:
1. If there are enough logs, we can simply slam the database trying to get the messages.
2. I believe `activity_log_id` isn't indexed by default so at some point these lookups are going to become quite slow if not timing out.

Reply:
Thanks @motnyk for the excellent input and @uncle-tee for the suggestion. I updated the code to implement it in a single join query: https://github.com/NangoHQ/nango/pull/633/commits/791a2ccf5049e265b3e253ddf3672fff9711e16a